File: android-usb-probe/app/src/main/java/com/gear66/heimdallusbprobe/MainActivity.java
package com.gear66.heimdallusbprobe;

import android.app.Activity;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.hardware.usb.UsbConstants;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbEndpoint;
import android.hardware.usb.UsbInterface;
import android.hardware.usb.UsbManager;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.Locale;

public final class MainActivity extends Activity {
    private static final String ACTION_USB_PERMISSION =
            "com.gear66.heimdallusbprobe.USB_PERMISSION";
    private static final int SAMSUNG_VENDOR_ID = 0x04E8;

    private UsbManager usbManager;
    private TextView logView;

    private final BroadcastReceiver permissionReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (!ACTION_USB_PERMISSION.equals(intent.getAction())) {
                return;
            }
            UsbDevice device = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE);
            boolean granted = intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false);
            append("Permission result: " + granted);
            if (granted && device != null) {
                probe(device);
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        usbManager = (UsbManager) getSystemService(Context.USB_SERVICE);

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(24, 24, 24, 24);

        TextView title = new TextView(this);
        title.setText("Heimdall Android USB Claim Probe");
        title.setTextSize(22f);
        title.setGravity(Gravity.CENTER_HORIZONTAL);
        root.addView(title, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT));

        Button runButton = new Button(this);
        runButton.setText("Run USB claim test");
        runButton.setOnClickListener(v -> startProbe());
        root.addView(runButton, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT));

        logView = new TextView(this);
        logView.setTextSize(15f);
        logView.setTextIsSelectable(true);
        logView.setMovementMethod(new ScrollingMovementMethod());
        root.addView(logView, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f));

        setContentView(root);

        IntentFilter filter = new IntentFilter(ACTION_USB_PERMISSION);
        registerReceiver(permissionReceiver, filter, Context.RECEIVER_NOT_EXPORTED);
        append("Ready. Put the tablet in Samsung Download Mode, connect OTG, then tap the button.");
    }

    @Override
    protected void onDestroy() {
        unregisterReceiver(permissionReceiver);
        super.onDestroy();
    }

    private void startProbe() {
        logView.setText("");
        append("Scanning USB devices...");

        UsbDevice samsung = null;
        for (UsbDevice device : usbManager.getDeviceList().values()) {
            append(String.format(Locale.US,
                    "Found: %s VID:PID=%04X:%04X interfaces=%d",
                    device.getDeviceName(), device.getVendorId(), device.getProductId(),
                    device.getInterfaceCount()));
            if (device.getVendorId() == SAMSUNG_VENDOR_ID) {
                samsung = device;
                break;
            }
        }

        if (samsung == null) {
            append("RESULT: No Samsung USB device found.");
            return;
        }

        if (usbManager.hasPermission(samsung)) {
            append("USB permission already granted.");
            probe(samsung);
            return;
        }

        append("Requesting Android USB permission...");
        Intent intent = new Intent(ACTION_USB_PERMISSION).setPackage(getPackageName());
        PendingIntent permissionIntent = PendingIntent.getBroadcast(
                this, 0, intent, PendingIntent.FLAG_MUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);
        usbManager.requestPermission(samsung, permissionIntent);
    }

    private void probe(UsbDevice device) {
        append(String.format(Locale.US, "Opening VID:PID=%04X:%04X...",
                device.getVendorId(), device.getProductId()));

        UsbDeviceConnection connection = usbManager.openDevice(device);
        if (connection == null) {
            append("RESULT: openDevice() returned null.");
            return;
        }

        boolean anyClaimed = false;
        try {
            for (int i = 0; i < device.getInterfaceCount(); i++) {
                UsbInterface intf = device.getInterface(i);
                append(String.format(Locale.US,
                        "Interface array=%d id=%d class=%02X subclass=%02X protocol=%02X endpoints=%d",
                        i, intf.getId(), intf.getInterfaceClass(), intf.getInterfaceSubclass(),
                        intf.getInterfaceProtocol(), intf.getEndpointCount()));

                for (int e = 0; e < intf.getEndpointCount(); e++) {
                    UsbEndpoint ep = intf.getEndpoint(e);
                    append(String.format(Locale.US,
                            "  endpoint[%d] address=%02X direction=%s type=%d maxPacket=%d",
                            e, ep.getAddress(),
                            ep.getDirection() == UsbConstants.USB_DIR_IN ? "IN" : "OUT",
                            ep.getType(), ep.getMaxPacketSize()));
                }

                boolean claimed;
                try {
                    claimed = connection.claimInterface(intf, true);
                } catch (RuntimeException ex) {
                    append("  claimInterface(force=true) threw: " + ex);
                    continue;
                }

                append("  claimInterface(force=true) = " + claimed);
                if (claimed) {
                    anyClaimed = true;
                    connection.releaseInterface(intf);
                    append("  releaseInterface() completed");
                }
            }

            append(anyClaimed
                    ? "RESULT: SUCCESS — Android native USB API claimed at least one interface."
                    : "RESULT: FAILURE — Android native USB API could not claim any interface.");
        } finally {
            connection.close();
            append("Connection closed. No firmware data was written.");
        }
    }

    private void append(String line) {
        logView.append(line + "\n");
    }
}

File: android-app/src/main/java/me/gear66/heimdallusb/MainActivity.kt
package me.gear66.heimdallusb

import android.app.Activity
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.hardware.usb.UsbConstants
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbDeviceConnection
import android.hardware.usb.UsbInterface
import android.hardware.usb.UsbManager
import android.os.Build
import android.os.Bundle
import android.view.Gravity
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView

class MainActivity : Activity() {
    private lateinit var usbManager: UsbManager
    private lateinit var status: TextView
    private var connection: UsbDeviceConnection? = null
    private var claimedInterface: UsbInterface? = null

    private val permissionAction by lazy { "$packageName.USB_PERMISSION" }

    private val permissionReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action != permissionAction) return

            val device = if (Build.VERSION.SDK_INT >= 33) {
                intent.getParcelableExtra(UsbManager.EXTRA_DEVICE, UsbDevice::class.java)
            } else {
                @Suppress("DEPRECATION")
                intent.getParcelableExtra(UsbManager.EXTRA_DEVICE)
            }

            if (intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false) && device != null) {
                openAndClaim(device)
            } else {
                append("USB permission denied.")
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        usbManager = getSystemService(USB_SERVICE) as UsbManager

        status = TextView(this).apply {
            textSize = 16f
            setPadding(24, 24, 24, 24)
            text = "Heimdall Android USB bridge diagnostic\n\n"
        }

        val scanButton = Button(this).apply {
            text = "Find and claim Samsung device"
            setOnClickListener { findSamsungDevice() }
        }

        val content = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER_HORIZONTAL
            setPadding(24, 36, 24, 24)
            addView(scanButton, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            addView(status, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        }

        setContentView(ScrollView(this).apply { addView(content) })

        val filter = IntentFilter(permissionAction)
        if (Build.VERSION.SDK_INT >= 33) {
            registerReceiver(permissionReceiver, filter, RECEIVER_NOT_EXPORTED)
        } else {
            @Suppress("DEPRECATION")
            registerReceiver(permissionReceiver, filter)
        }

        findSamsungDevice()
    }

    private fun findSamsungDevice() {
        releaseUsb()
        status.text = "Scanning attached USB devices...\n"

        val samsung = usbManager.deviceList.values.firstOrNull { it.vendorId == SAMSUNG_VENDOR_ID }
        if (samsung == null) {
            append("No Samsung USB device is attached.")
            append("Put the tablet in Download Mode and reconnect the cable.")
            return
        }

        append("Found ${hex(samsung.vendorId)}:${hex(samsung.productId)}")
        append("Device name: ${samsung.deviceName}")
        append("Interfaces reported: ${samsung.interfaceCount}")

        if (usbManager.hasPermission(samsung)) {
            append("Android USB permission already granted.")
            openAndClaim(samsung)
        } else {
            append("Requesting Android USB permission...")
            val pendingIntent = PendingIntent.getBroadcast(
                this,
                0,
                Intent(permissionAction).setPackage(packageName),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
            usbManager.requestPermission(samsung, pendingIntent)
        }
    }

    private fun openAndClaim(device: UsbDevice) {
        releaseUsb()

        val dataInterface = (0 until device.interfaceCount)
            .map { device.getInterface(it) }
            .firstOrNull {
                it.interfaceClass == UsbConstants.USB_CLASS_CDC_DATA &&
                    it.endpointCount >= 2
            }

        if (dataInterface == null) {
            append("FAILED: CDC data interface was not found.")
            describeInterfaces(device)
            return
        }

        append(
            "Selected interface ${dataInterface.id}: " +
                "class=0x${dataInterface.interfaceClass.toString(16)}, " +
                "endpoints=${dataInterface.endpointCount}"
        )

        val opened = usbManager.openDevice(device)
        if (opened == null) {
            append("FAILED: UsbManager.openDevice() returned null.")
            return
        }

        connection = opened
        append("UsbDeviceConnection opened. fd=${opened.fileDescriptor}")

        val claimed = opened.claimInterface(dataInterface, true)
        if (!claimed) {
            append("FAILED: Android UsbDeviceConnection.claimInterface() returned false.")
            releaseUsb()
            return
        }

        claimedInterface = dataInterface
        append("SUCCESS: Android claimed interface ${dataInterface.id}.")
        append("This proves the interface can be owned through Android's USB Host API.")
        append("Next engineering step: run Heimdall protocol transfers inside this app/JNI process.")
    }

    private fun describeInterfaces(device: UsbDevice) {
        for (index in 0 until device.interfaceCount) {
            val intf = device.getInterface(index)
            append(
                "Interface[$index] id=${intf.id} class=0x${intf.interfaceClass.toString(16)} " +
                    "subclass=0x${intf.interfaceSubclass.toString(16)} endpoints=${intf.endpointCount}"
            )
        }
    }

    private fun releaseUsb() {
        claimedInterface?.let { connection?.releaseInterface(it) }
        claimedInterface = null
        connection?.close()
        connection = null
    }

    private fun append(message: String) {
        status.append("$message\n")
    }

    private fun hex(value: Int): String = "0x%04X".format(value)

    override fun onDestroy() {
        releaseUsb()
        unregisterReceiver(permissionReceiver)
        super.onDestroy()
    }

    companion object {
        private const val SAMSUNG_VENDOR_ID = 0x04E8
    }
}
